clang/tools/libclang/CCAS.cpp

```diff
@@ -74,6 +138,332 @@ void clang_experimental_cas_Databases_dispose(CXCASDatabases CDBs) {
+      WComp.CachedResult.getOutput(OutputIdx);
+  return cxstring::createDup(WComp.CAS->getID(Output.Object).toString());
+}
+
+bool clang_experimental_cas_CachedCompilation_isOutputMaterialized(
+    CXCASCachedCompilation CComp, size_t OutputIdx) {
+  WrappedCachedCompilation &WComp = *unwrap(CComp);
+  CompileJobCacheResult::Output Output =
+      WComp.CachedResult.getOutput(OutputIdx);
+  bool IsMaterialized = false;
+  // FIXME: Propagate error to caller instead of calling `report_fatal_error`.
+  // In practice this should not fail because it checks the local CAS only.
+  if (Error E =
+          WComp.CAS->isMaterialized(Output.Object).moveInto(IsMaterialized))
+    llvm::report_fatal_error(std::move(E));
+  return IsMaterialized;
+}
+
+void clang_experimental_cas_CachedCompilation_makeGlobal(
+    CXCASCachedCompilation CComp, void *Ctx,
+    void (*Callback)(void *Ctx, CXError), CXCASCancellationToken *OutToken) {
+  if (OutToken)
+    *OutToken = nullptr;
+  WrappedCachedCompilation &WComp = *unwrap(CComp);
+  CompileJobCacheResult &CacheResult = WComp.CachedResult;
+  WComp.AC->putAsync(WComp.CacheKey, CacheResult.getID(), /*Globally=*/true,
+                     [Ctx, Callback](Error E) {
+                       Callback(Ctx, cxerror::create(std::move(E)));
+                     });
+}
+
+CXCASReplayResult clang_experimental_cas_replayCompilation(
+    CXCASCachedCompilation CComp, int argc, const char *const *argv,
+    const char *WorkingDirectory, void * /*reserved*/, CXError *OutError) {
+  WrappedCachedCompilation &WComp = *unwrap(CComp);
+
+  if (OutError)
+    *OutError = nullptr;
+
+  IntrusiveRefCntPtr<DiagnosticIDs> DiagID(new DiagnosticIDs());
+  IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts = new DiagnosticOptions();
+  SmallString<128> DiagsBuffer;
+  llvm::raw_svector_ostream DiagOS(DiagsBuffer);
+  auto *DiagsPrinter = new TextDiagnosticPrinter(DiagOS, DiagOpts.get());
+  DiagnosticsEngine Diags(DiagID, DiagOpts.get(), DiagsPrinter);
+
+  auto Invok = std::make_shared<CompilerInvocation>();
+  ArrayRef Argv(argv, argc);
+  bool Success = CompilerInvocation::CreateFromArgs(*Invok, Argv.drop_front(),
+                                                    Diags, Argv.front());
+  if (!Success) {
+    if (OutError)
+      *OutError = cxerror::create(DiagOS.str());
+    return nullptr;
+  }
+
+  StringRef WorkingDir = WorkingDirectory;
+  if (!WorkingDir.empty())
+    Invok->getFileSystemOpts().CASFileSystemWorkingDirectory = WorkingDir;
+
+  SmallString<256> DiagText;
+  std::optional<int> Ret;
+  if (Error E =
+          CompileJobCache::replayCachedResult(std::move(Invok), WComp.CacheKey,
+                                              WComp.CachedResult, DiagText)
+              .moveInto(Ret)) {
+    if (OutError)
+      *OutError = cxerror::create(std::move(E));
+    else
+      llvm::consumeError(std::move(E));
+    return nullptr;
+  }
+
+  if (!Ret)
+    return nullptr;
+  assert(*Ret == 0);
```

Comment:
Maybe add a string for why this must be 0 (I assume error should have been reported above otherwise)

Reply:
Done in https://github.com/apple/llvm-project/pull/7295/commits/1e1e184c4b2e51fe95fd619715d3a122018fb009